hal_st/stm32fxxx/UartStmDma.cpp

```diff
@@ -1,79 +1,66 @@
 
 namespace hal
 {
     namespace
     {
-        volatile void* PeripheralAddress(infra::MemoryRange<USART_TypeDef* const> table, uint8_t uartIndex)
+        volatile void* TransmitRegister(uint8_t uartIndex)
         {
 #if defined(USART_TDR_TDR)
-            return &table[uartIndex]->TDR;
+            return &peripheralUart[uartIndex]->TDR;
 #else
-            return &table[uartIndex]->DR;
+            return &peripheralUart[uartIndex]->DR;
 #endif
         }
     }
 
     UartStmDma::UartStmDma(DmaStm::TransmitStream& transmitStream, uint8_t oneBasedIndex, GpioPinStm& uartTx, GpioPinStm& uartRx, const Config& config)
-        : UartStmDma{ transmitStream, oneBasedIndex, uartTx, uartRx, hal::dummyPinStm, hal::dummyPinStm, config, false }
-    {}
+        : UartStm(oneBasedIndex, uartTx, uartRx, config)
+        , transmitDmaChannel{ transmitStream, TransmitRegister(oneBasedIndex - 1), 1, [this]
+            {
+                TransferComplete();
+            } }
+    {
+        peripheralUart[uartIndex]->CR3 |= USART_CR3_DMAT;
+    }
 
     UartStmDma::UartStmDma(DmaStm::TransmitStream& transmitStream, uint8_t oneBasedIndex, GpioPinStm& uartTx, GpioPinStm& uartRx, GpioPinStm& uartRts, GpioPinStm& uartCts, const Config& config)
-        : UartStmDma{ transmitStream, oneBasedIndex, uartTx, uartRx, uartRts, uartCts, config, true }
-    {}
-
-    UartStmDma::UartStmDma(DmaStm::TransmitStream& transmitStream, uint8_t oneBasedIndex, GpioPinStm& uartTx, GpioPinStm& uartRx, GpioPinStm& uartRts, GpioPinStm& uartCts, const Config& config, bool hasFlowControl)
-        : uartIndex(oneBasedIndex - 1)
-        , uartTx(uartTx, PinConfigTypeStm::uartTx, oneBasedIndex)
-        , uartRx(uartRx, PinConfigTypeStm::uartRx, oneBasedIndex)
-        , uartRts(uartRts, PinConfigTypeStm::uartRts, oneBasedIndex)
-        , uartCts(uartCts, PinConfigTypeStm::uartCts, oneBasedIndex)
-        , uartArray(peripheralUart)
-        , uartIrqArray(peripheralUartIrq)
-        , transmitDmaChannel{ transmitStream, PeripheralAddress(uartArray, uartIndex), 1, [this]()
+        : UartStm(oneBasedIndex, uartTx, uartRx, uartRts, uartCts, config)
+        , transmitDmaChannel{ transmitStream, TransmitRegister(oneBasedIndex - 1), 1, [this]
             {
                 TransferComplete();
             } }
     {
-        RegisterInterrupt(config);
-        EnableClockUart(uartIndex);
-        UartStmHalInit(config, hasFlowControl);
+        peripheralUart[uartIndex]->CR3 |= USART_CR3_DMAT;
     }
 
 #if defined(HAS_PERIPHERAL_LPUART)
     UartStmDma::UartStmDma(DmaStm::TransmitStream& transmitStream, uint8_t oneBasedIndex, GpioPinStm& uartTx, GpioPinStm& uartRx, LpUart lpUart, const Config& config)
-        : UartStmDma(transmitStream, oneBasedIndex, uartTx, uartRx, dummyPinStm, dummyPinStm, lpUart, config, false)
-    {}
+        : UartStm(oneBasedIndex, uartTx, uartRx, lpUart, config)
+        , transmitDmaChannel{ transmitStream, TransmitRegister(oneBasedIndex - 1), 1, [this]
```

Comment:
It should be possible to use `uartIndex` here instead of `oneBasedIndex - 1`. Also same for locations in UartStmDuplexDma.cpp

Reply:
Thanks for catching that! I forgot I made uartIndex protected just for this reason! Will make the change